src/httpClient.ts

```diff
@@ -11,26 +13,54 @@ export interface RequestParams {
 }
 export type RequestResponse<T> = Promise<{ body?: T; headers: Headers }>;
 
+export interface HTTPClient {
+  doRequest<T>(_: RequestParams): RequestResponse<T>;
+}
+
 /**
- * This a curried helper function that returns a function for making fetch
- * requests against the API.
+ * This a helper function that returns a class for making fetch requests
+ * against the API.
  *
  * @param baseUrl The base URL of the API endpoint.
  * @param apiKey The API key to use for authentication.
  *
- * @returns A function to make requests against the API.
+ * @returns An HTTPClient to make requests against the API.
  */
-export const createDoRequest =
-  <T>(baseUrl: string, apiKey: string) =>
-  async ({
+export const createHTTPClient = (baseUrl: string, apiKey: string) =>
+  new DefaultHTTPClient(baseUrl, apiKey);
+
+class DefaultHTTPClient implements HTTPClient {
+  private agent: Agent;
+  private baseUrl: string;
+  private apiKey: string;
+  readonly userAgent = `tpuf-typescript/${version}`;
+
+  constructor(baseUrl: string, apiKey: string) {
+    this.baseUrl = baseUrl;
+    this.apiKey = apiKey;
+
+    this.agent = new Agent({
+      keepAliveTimeout: 1 * 60 * 60 * 1000, // 1 hour, default duration socket can be kept alive
+      keepAliveMaxTimeout: 24 * 60 * 60 * 1000, // 24 hours, maximum configurable timeout with server hint
+    });
+
+    // warm up a single connection
+    void fetch(this.baseUrl, {
+      method: "HEAD",
+      headers: { "User-Agent": this.userAgent },
+      dispatcher: this.agent,
+    });
```

Comment:
not sure if we really want to do this

Reply:
I think we have to do it, and we could make it a configuration later... don't you technically have to do this in a `setInterval` in relation to the keepalive?

Reply:
The idea is to always set your idle timeout greater than your expected request interval, so it keeps itself alive. And `warmConnections` is just to avoid paying the connection cost for your first request.